### PR TITLE
Dtspo 440 automation cert

### PR DESCRIPTION
**Pull request by msl8r**
### Change description ###
Added code to:
  - Create a service principal that the automation account can use to authenticate
  - Create a certificate for the service principal that is then used by the runbook
  - Configure the Automation account connection using the service principal and cert

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: terraform/providers.tf [hunk 1]
```diff
@@ -1,13 +1,23 @@
...
 }
 provider "azurerm" {
+  subscription_id      = "1baf5470-1c3e-40d3-a6f7-74bfbce4b348"
+  features {}
+}
+
+provider "azuread" {
+}
+
+provider "azurerm" {
```
Comment (msl8r) [inline terraform/providers.tf:19]: need to leave this here until an apply is run

File: terraform/runbook.tf [hunk 2]
```diff
@@ -45,5 +57,24 @@ resource "azurerm_automation_module" "az_keyvault_module" {
...
   # This module cannot be imported unless the Accounts Module has been imported first
-  depends_on = [ azurerm_automation_module.az_accounts_module ]
+  depends_on = [azurerm_automation_module.az_accounts_module]
+}
+
+resource "azurerm_automation_certificate" "github_membership_runas_cert" {
+  name                    = "AzureRunAsCertificate"
+  resource_group_name     = azurerm_automation_account.github_membership_automation.resource_group_name
+  automation_account_name = azurerm_automation_account.github_membership_automation.name
+  base64                  = data.azurerm_key_vault_secret.runas_cert_secret.value
```
Comment (msl8r) [inline terraform/runbook.tf:67]: Had to get the cert using a azurerm_key_vault_secret  data source to get the full pfx version

File: terraform/service-principal.tf [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  service_principal_id = azuread_service_principal.automation_account.id
+  type                 = "AsymmetricX509Cert"
+  encoding             = "hex"
+  value                = azurerm_key_vault_certificate.automation_account.certificate_data
+  end_date             = azurerm_key_vault_certificate.automation_account.certificate_attribute[0].expires
+}
+
+resource "azurerm_role_assignment" "key_vault" {
+  scope                = data.azurerm_key_vault.cert_key_vault.id
+  role_definition_name = "Key Vault Secrets User"
```
Comment (timja) [inline terraform/service-principal.tf:30]: does this mean you had to swap the access policy method to role from access policy on this vault

File: terraform/tls-cert.tf [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+      reuse_key  = true
+    }
+
+    lifetime_action {
+      action {
+        action_type = "AutoRenew"
+      }
+
+      trigger {
+        days_before_expiry = 30
```
Comment (timja) [inline terraform/tls-cert.tf:23]: nice but i assume the automation account needs updating too when it expires?

File: terraform/service-principal.tf (hunk 3 above)
Comment (msl8r) [reply terraform/service-principal.tf:30]: No reason for using the role assignment, access policy slipped my mind. Is the access policy method preferred, I can change it if so.

File: terraform/tls-cert.tf (hunk 4 above)
Comment (msl8r) [reply terraform/tls-cert.tf:23]: Yeah I think so, haven't tested though. We'd need to apply tf again or have something like a runbook to update the automation account and sp

File: terraform/service-principal.tf (hunk 3 above)
Comment (timja) [reply terraform/service-principal.tf:30]: the role assignment method is preview and as far as I know you can't mix and match the two, we use access policies everywhere

File: terraform/tls-cert.tf (hunk 4 above)
Comment (timja) [reply terraform/tls-cert.tf:23]: thought so

File: terraform/service-principal.tf (hunk 3 above)
Comment (msl8r) [reply terraform/service-principal.tf]: done

**Review by timja — APPROVED** [review]

**Pull request merged**